src/Sylius/Bundle/ShippingBundle/Form/EventListener/AddCodeFromSubscriber.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\ShippingBundle\Form\EventListener;
+
+use Sylius\Component\Resource\Model\CodeAwareInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\Intl\Exception\UnexpectedTypeException;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+class AddCodeFromSubscriber implements EventSubscriberInterface
+{
+
+    /**
+     * @return array
+     */
+    public static function getSubscribedEvents()
+    {
+        return array(
+            FormEvents::PRE_SET_DATA => 'preSetData',
+        );
+    }
+
+    /**
+     * @param FormEvent $event
+     */
+    public function preSetData(FormEvent $event)
+    {
+        $entityWithImmutableCode= $event->getData();
+        $form = $event->getForm();
+
+        if (!$entityWithImmutableCode instanceof CodeAwareInterface) {
+            throw new UnexpectedTypeException('Expected type was: ', CodeAwareInterface::class);
```

Comment:
Use `Sylius\Component\Resource\Exception\UnexpectedTypeException`, pass object and interface name it will generate approriate message